### PR TITLE
[test] test/main.lua enhancements

### PR DESCRIPTION
**Pull request by myk002**
#1690

new test/main API is backwards compatible with existing usage.

- allow configuration set in test_config.json to be dynamically overridden on the commandline
- make `expect.table_eq()` diff recursively (all existing tests that use it still pass)
- fix implementation of testing for multiple test name filters. the original implementation filtered out too many tests if more than one filter was specified.
- add ability to filter by mode (title, none, etc.)
- add ability to rerun tests that have already been run (I'm actually not sure what the purpose of `test_status.json` is. why would we want to skip a test if it has already been run? If you're running the tests again, haven't you just edited the code and wouldn't you want to see if tests that previously passed are now failing?)
- done command can now be passed as space-separated tokens or as one quoted string
- added output text that says what your test dir is and what your filters are
- added tests for new table_eq implementation

### CODE REVIEW

File: test/main.lua [hunk 1]
```diff
@@ -162,6 +245,7 @@ function build_test_env()
             mode = 'none',
         },
         expect = {},
+        expect_raw = expect,
```
Comment (lethosor) [inline test/main.lua:248]: I would suggest importing `test/main.lua` from the test instead, since `expect` is a global already. It would need to be [tweaked slightly](https://docs.dfhack.org/en/stable/docs/Lua%20API.html#importing-scripts) to only call `main()` when invoked as a command, but I think it should work when that's done.

Comment (myk002) [reply test/main.lua]: good idea. I rewrote main.lua to be importable.

**Comment by BenLubar** [comment]
If you're going to pass the done_command as an array, you need to also update https://github.com/DFHack/dfhack/blob/develop/travis/run-tests.py#L76 to generate it as two arguments rather than one argument with a space

**Comment by myk002** [comment]
done. thanks!

File: test/test_test.lua [hunk 2]
```diff
@@ -0,0 +1,36 @@
+local expect_raw = reqscript('test/main').expect
```
Comment (lethosor) [inline test/test_test.lua:1]: > rename test.lua to test_test.lua since I plan to rename main.lua to test.lua in a future PR

I think we should reconsider where the test runner lives in the future, in that case. I didn't really think through where the runner should live, but it shouldn't be in the same namespace as the tests themselves. Essentially, my proposal is to:
* move test/main.lua to somewhere else in the DFHack repo and install it to hack/scripts if BUILD_TESTS is enabled. (Now that I think about it, it probably lives at `test/main` because I couldn't think of another place to put it in the DFHack repo, but maybe the `ci` folder could work.)
* don't add the `test` folder to the script path, since none of the test cases themselves are intended to be run directly (and they wouldn't work if they were). This would allow tests to have the same names as other scripts.

This is something I can handle if it's not too disruptive to your test work. Just wanted to make sure our plans were aligned - feel free to suggest something else.

Comment (myk002) [reply test/test_test.lua:1]: that sounds good to me. Changes to the test runner don't conflict with anything I have in the pipeline. I'm just writing unit tests ATM.

Now that we're going to install the test runner directly into the `scripts/` directory, I do suggest still renaming `main.lua` to `test.lua`. Running tests with `test/main` made sense. Running tests with just `main` isn't quite as intuitive.

Comment (lethosor) [reply test/test_test.lua:1]: Oh, definitely - forgot to mention it, but I was thinking the same thing.

**Comment by lethosor** [comment]
> * add ability to rerun tests that have already been run (I'm actually not sure what the purpose of `test_status.json` is. why would we want to skip a test if it has already been run? If you're running the tests again, haven't you just edited the code and wouldn't you want to see if tests that previously passed are now failing?)

I missed this earlier. The purpose is to allow DF to be restarted and resume tests where they left off if a single test causes DF to crash. This is somewhat more likely in our case than in other projects because we have less control over the environment that tests are running in.

File: test/main.lua [hunk 3]
```diff
@@ -292,55 +379,75 @@ function run_test(test, status, counts)
...
+
+    local status = {}
+    if not config.nocache then
+        status = load_test_status() or status
+        for i = #tests, 1, -1 do
+            local test = tests[i]
+            if not status[test.full_name] then
+                status[test.full_name] = TestStatus.PENDING
+            elseif status[test.full_name] ~= TestStatus.PENDING then
+                print(('skipping test: %s: state = %s)'):format(
```
Comment (lethosor) [inline test/main.lua:438]: Heh, thanks for preserving the `)`, but that's a typo from e6c9db0a18a49ab3777d904151a40bb1d660a398 :)

File: test/main.lua [hunk 4]
```diff
@@ -354,6 +461,55 @@ function main()
...
+        })
+
+    if help then print(help_text) return end
+
+    local done_command = table.concat(other_args, ' ')
+    local config = load_test_config(CONFIG_FILE)
+
+    -- override config with any params specified on the commandline
+    if test_dir then config.test_dir = test_dir end
+    if nocache then config.nocache = true end
```
Comment (lethosor) [inline test/main.lua:485]: Honestly, I might end up inverting this. CI always runs tests through `run-tests.py`, which could pass a `--cache` flag to `test/main.lua` fairly easily and wouldn't need CI changes.

**Review by lethosor — APPROVED** [review]

File: test/main.lua (hunk 3 above)
Comment (myk002) [reply test/main.lua:438]: haha, I noticed, but I thought it was just stylized formatting -- it actually looks reasonable in the output

File: test/main.lua (hunk 4 above)
Comment (myk002) [reply test/main.lua:485]: that works. maybe rename `-n, --nocache` to `-r, --resume`. Want me to do that?

Comment (lethosor) [reply test/main.lua:485]: Makes sense to me. Feel free
